Added location option to test target

KrisRJack opened the pull request:
Resolves #1150

KrisRJack commented on Sources/XcodeGenKit/SchemeGenerator.swift:
```diff
@@ -187,12 +187,24 @@ public class SchemeGenerator {
                 randomExecutionOrdering: testTarget.randomExecutionOrder,
-                buildableReference: testBuilEntries.buildableReference,
+                buildableReference: testBuildEntries.buildableReference,
+                locationScenarioReference: locationScenarioReference,
```
Created PR for this here in XcodeProj: https://github.com/tuist/XcodeProj/pull/654

KrisRJack commented:
Hey All! Let me know if there is anymore that I can do here on my end to help move this PR forward. Thank you! (cc @yonaskolb)

farshadtx commented:
@yonaskolb can we get a little love to this one?!

LuAndreCast commented:
@yonaskolb can we get this fixed please?

yonaskolb commented on Sources/ProjectSpec/Scheme.swift:
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
 
         public struct TestTarget: Equatable, ExpressibleByStringLiteral {
+            
+            public struct SimulateLocation: Equatable {
```
Let's rename this to `Location`

yonaskolb commented on Sources/ProjectSpec/Scheme.swift:
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
+                    case gpx = "0"
+                }
+                
+                public var location: String?
```
Any reason this needs to be optional?

yonaskolb commented on Sources/ProjectSpec/Scheme.swift:
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
+                    return .predefined
+                }
+                
+                public init(location: String) {
```
```suggestion
                public init(_ location: String) {
```

yonaskolb commented on Sources/ProjectSpec/Scheme.swift:
```diff
@@ -559,6 +593,9 @@ extension Scheme.Test.TestTarget: JSONEncodable {
             dict["parallelizable"] = parallelizable
         }
+        if let location = location {
+            dict["location"] = location
```
This just puts the struct into the dictionary. Write the string
```suggestion
            dict["location"] = location.location
```

yonaskolb commented on Sources/ProjectSpec/Scheme.swift:
```diff
@@ -536,6 +569,7 @@ extension Scheme.Test.TestTarget: JSONObjectConvertible {
         targetReference = try TargetReference(jsonDictionary.json(atKeyPath: "name"))
         randomExecutionOrder = jsonDictionary.json(atKeyPath: "randomExecutionOrder") ?? Scheme.Test.TestTarget.randomExecutionOrderDefault
         parallelizable = jsonDictionary.json(atKeyPath: "parallelizable") ?? Scheme.Test.TestTarget.parallelizableDefault
+        location = jsonDictionary.json(atKeyPath: "location") ?? nil
```
Decode the location as a string. The spec says it's a string. At the moment you would need:
```
location:
  location: abc.gpx
```

yonaskolb commented on Docs/ProjectSpec.md:
```diff
@@ -834,6 +834,7 @@ A multiline script can be written using the various YAML multiline methods, for
 - [x] **name**: **String** - The name of the target
 - [ ] **parallelizable**: **Bool** - Whether to run tests in parallel. Defaults to false
 - [ ] **randomExecutionOrder**: **Bool** - Whether to run tests in a random order. Defaults to false
+- [ ] **location**: **String** - GPX file or predefined value for simulating location. See [Simulate Location](#simulate-location).
```
```suggestion
- [ ] **location**: **String** - GPX file or predefined value for simulating location. See [Simulate Location](#simulate-location) for location examples.
```

yonaskolb reviewed (commented):
Thanks for this @KrisRJack. Left some comments, mainly around simplifying the location spec to a string instead of an object.
Can you also:
- add a decoding test to SpecLoadingTests
- add a generator test to SchemeGeneratorTests
- add an example to the Fixtures/TestProject/project.yml

KrisRJack commented:
Hey Yonas! Thank you so much for the feedback! I'll have these changes in shortly.

KrisRJack commented:
Hey @yonaskolb! Thanks again for the feedback and the review. Below, I've outlined the changes made to this PR:

- I removed the `SimulateLocation` struct entirely and just passed in a string of the location 
- Removed `SimulateLocation` from SchemeGenerator.swift and gathered location reference type and identifier from given string
- Expanded upon existing tests in SpecLoadingTests and SchemeGeneratorTests to include location in test targets. The commits test both for custom `.gpx` files and default locations such as "New York, NY, USA" 
- Added an example to the Fixtures/TestProject/project.yml

Thanks again and let me know if there's anything more that I can add to this.

yonaskolb reviewed (approved):
Fantastic! Thanks again!

KrisRJack commented:
Hey @yonaskolb! Thanks again for your review and merging! Really awesome to see this in the project. 

Any chance you can update the release to 2.26.0 as well so our team can apply this feature to our repo? Thanks again!

yonaskolb commented:
@KrisRJack released 👍
